object.go

```diff
@@ -26,13 +39,83 @@ import (
 // and AddChildObjects to ensure that the correct types are used for values.
 type Object map[string]any
 
+// Duration is the value of an attribute of type duration.
+// It is the sum of all the fields' durations.
+type Duration struct {
+	// Nanos is a duration as a number of nanoseconds.
+	Nanos int32 `json:"nanos,omitempty"`
+
+	// Seconds is a duration as a number of seconds.
+	Seconds int64 `json:"seconds,omitempty"`
+
+	// Days is a duration as a number of days.
+	Days int64 `json:"days,omitempty"`
+
+	// Months is a duration as a number of months.
+	Months int64 `json:"months,omitempty"`
+}
+
+// ParseDuration parses a valid ISO8601 duration string into a Duration.
+// Years as duration is not supported.
+// Fractional components of durations are not supported.
+func ParseISO8601Duration(durationStr string) (*Duration, error) {
+	d, err := duration.Parse(durationStr)
+	if err != nil {
+		return nil, errors.New("failed to parse the duration string: " + durationStr)
+	}
+
+	if d.Years != 0 {
+		return nil, errors.New("years as duration is not supported")
+	}
+
+	if err := hasFractionalComponent(d); err != nil {
+		return nil, err
+	}
+
+	// Express minutes and hours as seconds.
+	d.Seconds += d.Minutes * float64(Minute)
+	d.Seconds += d.Hours * float64(Hour)
+
+	// Express weeks as days.
+	d.Days += d.Weeks * float64(Week)
+
+	return &Duration{Seconds: int64(d.Seconds), Days: int64(d.Days), Months: int64(d.Months)}, nil
```

Comment:
1. Fractional values should be supported. There's no technical limitation preventing us from doing so.
2. It's not OK to not properly parse sub-second durations.
```suggestion
	// Convert years into months, weeks into days, and minutes and hours into seconds.
	d.Months += d.Years * 12.0
	d.Days += d.Weeks * 7.0
	d.Seconds += d.Hours * 3_600.0
	d.Seconds += d.Minutes * 60.0

	// Round the numbers of months, days, seconds, and nanoseconds.
	months, monthsFraction := math.Modf(d.Months)
	d.Days += monthsFraction * 30.0 # Approximate 1 month = 30 days

	days, daysFraction := math.Modf(d.Days)
	d.Seconds += daysFraction * 24.0 * 3_600.0 # Approximate 1 day = 24*3600 seconds

	seconds, secondsFraction := math.Modf(d.Seconds)
	
	nanos := secondsFraction * 1_000_000_000.0

	return &Duration{
		Months: int64(months),
		Days: int64(days),
		Seconds: int64(seconds),
		Nanos: int64(nanos),
	}, nil
```

Reply:
I was concerned about the inaccuracies that might be caused by the approximations we need to perform for fractional values e.g. `1 month = 30 days`

Reply:
Yeah, but then that prevents you from having subsecond duraitons, which could be a big deal.
I would prefer to approximate than preventing that.

Reply:
Actually, we should check in ISO 8601 how they handle fractional months...

Reply:
In ISO 8601 Section 4.3.1:

>  In certain applications a month is regarded as a unit of time of 30 days.

So approximating 1 month = 30 days is OK IMHO.